src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMirrorHandler.java

```diff
@@ -0,0 +1,145 @@
+
+    BitbucketMirrorHandler(
+            BitbucketPluginConfiguration bitbucketPluginConfiguration,
+            BitbucketClientFactoryProvider bitbucketClientFactoryProvider,
+            JenkinsToBitbucketCredentials jenkinsToBitbucketCredentials,
+            BitbucketRepoFetcher bitbucketRepoFetcher) {
+        this.bitbucketPluginConfiguration = bitbucketPluginConfiguration;
+        this.bitbucketClientFactoryProvider = bitbucketClientFactoryProvider;
+        this.jenkinsToBitbucketCredentials = jenkinsToBitbucketCredentials;
+        this.bitbucketRepoFetcher = bitbucketRepoFetcher;
+    }
+
+    public EnrichedBitbucketMirroredRepository fetchRepository(MirrorFetchRequest mirrorFetchRequest) {
+        return this.fetchRepositories(mirrorFetchRequest)
+                .stream()
+                .filter(r -> r.getMirroringDetails().getMirrorName().equals(mirrorFetchRequest.getExistingMirrorSelection()))
+                .findFirst()
+                .orElseThrow(() -> new MirrorFetchException(
+                        "Unable to find the mirror" + mirrorFetchRequest.getExistingMirrorSelection()));
+    }
+
+    public StandardListBoxModel fetchAsListBox(MirrorFetchRequest mirrorFetchRequest) {
+        StandardListBoxModel options = new StandardListBoxModel();
+        final Option defaultSelected = new Option(DEFAULT_UPSTREAM_SERVER, "", true);
+        if (isEmpty(mirrorFetchRequest.getServerId()) ||
+            isEmpty(mirrorFetchRequest.getProjectNameOrKey()) ||
+            isEmpty(mirrorFetchRequest.getRepoNameOrSlug())) {
+            options.add(defaultSelected);
+            return options;
+        }
+
+        String existingSelection = mirrorFetchRequest.getExistingMirrorSelection();
+        List<EnrichedBitbucketMirroredRepository> repositories =
+                fetchRepositoriesQuietly(mirrorFetchRequest);
+        if (repositories.isEmpty()) {
+            options.add(defaultSelected);
+            return options;
+        }
+
+        List<Option> mirrors = repositories
+                .stream()
+                .map(mirroredRepo -> createOption(existingSelection, mirroredRepo))
+                .collect(Collectors.toList());
+        boolean isPresent = mirrors
+                .stream()
+                .anyMatch(option -> option.selected);
+        if (isPresent) {
+            options.add(new Option(DEFAULT_UPSTREAM_SERVER, ""));
+        } else {
+            options.add(new Option(DEFAULT_UPSTREAM_SERVER, "", true));
+        }
+        options.addAll(mirrors);
+        return options;
+    }
+
+    private List<EnrichedBitbucketMirroredRepository> fetchRepositories(MirrorFetchRequest mirrorFetchRequest) {
+        BitbucketServerConfiguration serverConfiguration =
+                bitbucketPluginConfiguration.getServerById(mirrorFetchRequest.getServerId())
+                        .orElseThrow(() -> new MirrorFetchException("Server config not found"));
```

Comment:
I don't think `MirrorFetchException` is appropriate here. Nothing went wrong with fetching the mirror itself - the arguments provided were incorrect.
Maybe `MirrorFetchRequest` can have a `BitbucketServerConfiguration` on it instead of a `serverId` and this handling can be done in the BitbucketSCM

Reply:
This is one of the use cases where throw will never be executed since the server id checking is already done. It can't be `BitbucketServerConfiguration` as when fetching the List options, I am returning Primary server if server id is empty.